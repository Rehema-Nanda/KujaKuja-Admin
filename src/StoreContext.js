// https://mobx-react.js.org/recipes-context

import React from "react";
import PropTypes from "prop-types";
import RootStore from "./stores/RootStore";

export const StoreContext = React.createContext(null);

export const StoreProvider = ({ children }) => {
    const store = new RootStore();
    return <StoreContext.Provider value={store}>{children}</StoreContext.Provider>;
};
StoreProvider.propTypes = {
    children: PropTypes.node.isRequired,
};

export const StoreConsumer = StoreContext.Consumer;

export const useStore = () => {
    const store = React.useContext(StoreContext);
    if (!store) {
        // this is especially useful in TypeScript so you don't need to be checking for null all the time
        throw new Error("useStore must be used within a StoreProvider.");
    }
    return store;
};
