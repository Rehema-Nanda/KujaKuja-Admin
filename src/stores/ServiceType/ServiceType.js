export default class ServiceType {
    store = null;
    id = null;
    name = null;
    createdAt = null;
    updatedAt = null;

    constructor(store, json = null) {
        this.store = store;

        if (json) {
            this.id = json.id;
            this.updateFromJson(json);
        }
    }

    get asJson() {
        return {
            name: this.name,
            created_at: this.createdAt,
            update_at: this.updatedAt,
        };
    }

    updateFromJson(json) {
        this.name = json.name;
        this.createdAt = json.created_at;
        this.updatedAt = json.updated_at;
    }

    async delete() {
        if (!this.id) {
            return;
        }
        await this.store.apiService.deleteServiceType(this.id);
    }

    async save() {
        const json = this.asJson;
        if (this.id) {
            await this.store.apiService.updateServiceType(this.id, json);
        }
        else {
            const responseData = await this.store.apiService.createServiceType(json);
            this.id = responseData.id;
        }
    }
}
