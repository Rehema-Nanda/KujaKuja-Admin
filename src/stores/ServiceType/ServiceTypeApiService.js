import ApiService from "../ApiService";

export default class ServiceTypeApiService extends ApiService {
    fetchAllServiceTypes = async (config) => {
        const serviceTypesResponse = await this.makeHttpCall("service_types/paginator", "post", config);
        return serviceTypesResponse.data; // TODO: change data structure returned by API
    };

    fetchServiceTypeById = async (id) => {
        const serviceTypeResponse = await this.makeHttpCall(`service_types/${id}`);
        return serviceTypeResponse.data.data; // TODO: change data structure returned by API
    };

    deleteServiceType = async (id) => {
        const deleteServiceTypeResponse = await this.makeHttpCall(`service_types/${id}`, "delete");
        // TODO: POST/PUT/DELETE endpoints should return string ID to be consistent with GET
        deleteServiceTypeResponse.data.id = deleteServiceTypeResponse.data.id.toString();
        return deleteServiceTypeResponse.data;
    };

    createServiceType = async (serviceTypeJson) => {
        const config = { data: serviceTypeJson };
        const createServiceTypeResponse = await this.makeHttpCall("service_types", "post", config);
        // TODO: POST/PUT/DELETE endpoints should return string ID to be consistent with GET
        createServiceTypeResponse.data.id = createServiceTypeResponse.data.id.toString();
        return createServiceTypeResponse.data;
    };

    updateServiceType = async (id, serviceTypeJson) => {
        const config = { data: serviceTypeJson };
        const updateServiceTypeResponse = await this.makeHttpCall(`service_types/${id}`, "put", config);
        // TODO: POST/PUT/DELETE endpoints should return string ID to be consistent with GET
        updateServiceTypeResponse.data.id = updateServiceTypeResponse.data.id.toString();
        return updateServiceTypeResponse.data;
    };
}
