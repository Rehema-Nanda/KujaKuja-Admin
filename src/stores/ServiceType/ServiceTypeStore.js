import {
    flow, decorate, observable, computed,
} from "mobx";
import ServiceType from "./ServiceType";
import ServiceTypeApiService from "./ServiceTypeApiService";
import AppConfig from "../../AppConfig";

class ServiceTypeStore {
    serviceTypes = [];
    isLoading = false;
    count = 0;

    constructor(rootStore) {
        this.rootStore = rootStore;
        this.apiService = new ServiceTypeApiService(AppConfig.API_URL);
        this.loadServiceTypes();
    }

    get serviceTypesSortedByName() {
        return this.serviceTypes.sort((a, b) => {
            if (a.name < b.name) {
                return -1;
            }
            else if (a.name > b.name) {
                return 1;
            }
            return 0;
        });
    }

    loadServiceTypes = flow(function* (config) { // eslint-disable-line func-names
        this.isLoading = true;
        try {
            const serviceTypesJson = yield this.apiService.fetchAllServiceTypes(config);
            this.serviceTypes.replace(serviceTypesJson.data.map((serviceTypeJson) => {
                return new ServiceType(
                    this,
                    serviceTypeJson,
                );
            }));
            this.count = serviceTypesJson.count;
            this.isLoading = false;
        }
        catch (error) {
            console.error(`ServiceTypeStore failed to load service types. ${error.message}`); // eslint-disable-line no-console
            this.isLoading = false;
        }
    });

    updateServiceTypeFromServer = flow(function* (id) { // eslint-disable-line func-names
        this.isLoading = true;
        const serviceTypesJson = yield this.apiService.fetchServiceTypeById(id);
        let serviceType = this.serviceTypes.find((st) => {
            return st.id === id;
        });

        if (serviceTypesJson.length === 0 && serviceType) {
            // service type no longer exists on the server, remove it locally
            yield this.removeServiceType(serviceType);
            serviceType = null;
        }
        if (serviceTypesJson.length > 0) {
            if (serviceType) {
                serviceType.updateFromJson(serviceTypesJson[0]);
            }
            else {
                serviceType = new ServiceType(this, serviceTypesJson[0]);
                this.serviceTypes.push(serviceType);
            }
        }
        this.isLoading = false;
        return serviceType;
    });

    removeServiceType = flow(function* (serviceType) { // eslint-disable-line func-names, require-yield
        this.serviceTypes.splice(this.serviceTypes.indexOf(serviceType), 1);
    });

    addServiceType = flow(function* (serviceType) { // eslint-disable-line func-names, require-yield
        this.serviceTypes.push(serviceType);
    });
}
decorate(ServiceTypeStore, {
    serviceTypes: observable,
    serviceTypesSortedByName: computed,
    isLoading: observable,
});

export default ServiceTypeStore;
