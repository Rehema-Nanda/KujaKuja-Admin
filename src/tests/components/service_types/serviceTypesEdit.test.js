import React from "react";
import { mount } from "enzyme";
import ServiceTypesEdit from "../../../components/service_types/edit";
import { StoreProvider } from "../../../StoreContext";
import ServiceTypeStore from "../../../stores/ServiceType/ServiceTypeStore";
import ServiceType from "../../../stores/ServiceType/ServiceType";

jest.mock("../../../stores/ServiceType/ServiceTypeStore");

const serviceTypes = [
    new ServiceType(
        null,
        {
            id: "1",
            name: "Test Service Type",
        },
    ),
];

const mockServiceTypeDelete = jest.fn();
serviceTypes[0].delete = mockServiceTypeDelete;

const mockUpdateServiceTypeFromServer = jest.fn(() => {
    return serviceTypes[0];
});
ServiceTypeStore.mockImplementation(() => {
    return {
        serviceTypes: serviceTypes,
        updateServiceTypeFromServer: mockUpdateServiceTypeFromServer,
    };
});

describe("ServiceTypesEdit component", () => {
    const testProps = {
        history: {
            push: jest.fn(),
        },
        match: {
            params: {
                id: 1,
            },
        },
    };
    const wrapper = mount(<ServiceTypesEdit {...testProps} />, {
        wrappingComponent: StoreProvider,
    });

    it("should render correctly", () => {
        expect(mockUpdateServiceTypeFromServer).toHaveBeenCalled();
        wrapper.update();
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("should push history when cancelForm function is called", () => {
        const e = { preventDefault: jest.fn() };
        const cancelForm = jest.spyOn(wrapper.instance(), "cancelForm");
        cancelForm(e);
        expect(testProps.history.push).toHaveBeenCalledWith("/service_types");
    });

    it("should change modal state when toggleDeleteConfirmationModal function is called", () => {
        const modalState = wrapper.instance().simpleState.isDeleteConfirmationModalOpen;
        const toggleDeleteConfirmationModal = jest.spyOn(wrapper.instance(), "toggleDeleteConfirmationModal");
        toggleDeleteConfirmationModal();
        expect(wrapper.instance().simpleState.isDeleteConfirmationModalOpen).toEqual(!modalState);
    });

    it("should redirect to list page upon element delete", async () => {
        const deleteFn = jest.spyOn(wrapper.instance(), "delete");
        await deleteFn();
        expect(mockServiceTypeDelete).toHaveBeenCalled();
        expect(testProps.history.push).toHaveBeenCalledWith("/service_types");
    });
});
