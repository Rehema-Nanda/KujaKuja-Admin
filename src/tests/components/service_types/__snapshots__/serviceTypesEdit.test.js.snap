// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceTypesEdit component should render correctly 1`] = `
"<ServiceTypesEdit history={{...}} match={{...}}>
  <div className=\\"page\\">
    <div className=\\"title-background-parent\\">
      <div className=\\"title-background-shape\\" />
    </div>
    <Banner pageTitle=\\"service type\\" elementAction=\\"update\\" buttonFunction={[Function]} buttonDisabled={false}>
      <Container tag=\\"div\\">
        <div className=\\"container\\">
          <Row tag=\\"div\\">
            <div className=\\"row\\">
              <Col sm={{...}} md={{...}} tag=\\"div\\" widths={{...}}>
                <div className=\\"col-sm-7 col-md-8\\">
                  <div className=\\"page-title\\">
                    <h1 className=\\"left\\">
                      Service type
                    </h1>
                    <p>
                      Manage your
                       
                      service type
                    </p>
                  </div>
                </div>
              </Col>
              <Col sm={{...}} md={{...}} tag=\\"div\\" widths={{...}}>
                <div className=\\"col-sm-5 col-md-4\\">
                  <div className=\\"page-title\\" />
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </Banner>
    <ConfirmationModal buttonFunction={[Function]} toggle={[Function]} isOpen={false}>
      <Modal isOpen={false} toggle={[Function]} autoFocus={true} centered={false} role=\\"dialog\\" backdrop={true} keyboard={true} zIndex={1050} fade={true} onOpened={[Function: noop]} onClosed={[Function: noop]} modalTransition={{...}} backdropTransition={{...}} />
    </ConfirmationModal>
    <Container tag=\\"div\\">
      <div className=\\"container\\">
        <div className=\\"white-page-bg\\">
          <Row tag=\\"div\\">
            <div className=\\"row\\">
              <Col lg={{...}} tag=\\"div\\" widths={{...}}>
                <div className=\\"col-lg-10 offset-lg-1\\">
                  <div>
                    <AvForm onValidSubmit={[Function]} tag={[Function: Form]} model={{...}} validationEvent={{...}} method=\\"get\\" onSubmit={[Function: onSubmit]} onKeyDown={[Function: onKeyDown]} onInvalidSubmit={[Function: onInvalidSubmit]}>
                      <Form noValidate={true} action=\\"#\\" method=\\"get\\" onSubmit={[Function]} onKeyDown={[Function: onKeyDown]} className=\\"av-valid\\" tag=\\"form\\">
                        <form noValidate={true} action=\\"#\\" method=\\"get\\" onSubmit={[Function]} onKeyDown={[Function: onKeyDown]} className=\\"av-valid\\">
                          <Row tag=\\"div\\">
                            <div className=\\"row\\">
                              <Col tag=\\"div\\" widths={{...}}>
                                <div className=\\"col\\">
                                  <AvField name=\\"name\\" label=\\"Name\\" required={true} value=\\"Test Service Type\\">
                                    <AvGroup check={false} disabled={[undefined]} row={false}>
                                      <FormGroup className=\\"\\" check={false} disabled={[undefined]} row={false} tag=\\"div\\">
                                        <div className=\\"position-relative form-group\\">
                                          <Label for=\\"name\\" className={[undefined]} hidden={[undefined]} size={[undefined]} tag=\\"label\\" widths={{...}}>
                                            <label htmlFor=\\"name\\" className=\\"\\">
                                              Name
                                            </label>
                                          </Label>
                                          <AvInput id=\\"name\\" className={[undefined]} size={[undefined]} disabled={[undefined]} readOnly={[undefined]} name=\\"name\\" required={true} value=\\"Test Service Type\\" validationEvent=\\"\\" validate={{...}} trueValue={true} falseValue={false} tag={[Function: Input]}>
                                            <Input size={[undefined]} disabled={[undefined]} readOnly={[undefined]} name=\\"name\\" required={true} value=\\"Test Service Type\\" onKeyUp={[Function: bound onKeyUpHandler]} onBlur={[Function: bound onBlurHandler]} onInput={[Function: bound onInputHandler]} onFocus={[Function: bound onFocusHandler]} onChange={[Function: bound onChangeHandler]} className=\\"is-untouched is-pristine av-valid\\" id=\\"name\\" type=\\"text\\">
                                              <input size={[undefined]} disabled={[undefined]} readOnly={[undefined]} name=\\"name\\" required={true} value=\\"Test Service Type\\" onKeyUp={[Function: bound onKeyUpHandler]} onBlur={[Function: bound onBlurHandler]} onInput={[Function: bound onInputHandler]} onFocus={[Function: bound onFocusHandler]} onChange={[Function: bound onChangeHandler]} id=\\"name\\" type=\\"text\\" className=\\"is-untouched is-pristine av-valid form-control\\" />
                                            </Input>
                                          </AvInput>
                                        </div>
                                      </FormGroup>
                                    </AvGroup>
                                  </AvField>
                                </div>
                              </Col>
                            </div>
                          </Row>
                          <BtnRow buttonLabel=\\"Update\\" cancelForm={[Function]}>
                            <Row className=\\"btn-row\\" tag=\\"div\\">
                              <div className=\\"btn-row row\\">
                                <Col tag=\\"div\\" widths={{...}}>
                                  <div className=\\"col\\" />
                                </Col>
                                <Col tag=\\"div\\" widths={{...}}>
                                  <div className=\\"col\\">
                                    <Button color=\\"primary\\" className=\\"primary cancel\\" onClick={[Function]} tag=\\"button\\">
                                      <button type=\\"button\\" onClick={[Function: bound onClick]} className=\\"primary cancel btn btn-primary\\" aria-label={{...}}>
                                        Back
                                      </button>
                                    </Button>
                                  </div>
                                </Col>
                              </div>
                            </Row>
                          </BtnRow>
                        </form>
                      </Form>
                    </AvForm>
                  </div>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </div>
    </Container>
  </div>
</ServiceTypesEdit>"
`;
